### ISSUE TITLE
Time stamps not updating on rerun tasks

### ISSUE THREAD

**Issue by IanSudbery** [cgat-developers/ruffus#105]
I'm having a rather odd problem. I've only just noticed it, but a guess it could have been happening a white. It appears that ruffus is not detecting changes in timestamps for files that have changed while it is running. 

Take the simple example pipeline below:

```python
from cgatcore import pipeline as P
from ruffus import transform, formatter
import sys
from time import sleep

@transform("*.input.file",
           formatter(),
           "{basename[0]}.output1.file")
def step1(infile, outfile):

    P.touch_file(outfile)


@transform(step1,
           formatter(),
           "{basename[0]}.output2.file")
def step2(infile, outfile):

    sleep(10)
    P.touch_file(outfile)


@transform(step2,
           formatter(),
           "{basename[0]}.output3.file")
def step3(infile,
          outfile):

    sleep(10)
    P.touch_file(outfile)

def main(argv=None):
    if argv is None:
        argv = sys.argv
    P.main(argv)

if __name__ == "__main__":
    sys.exit(P.main(sys.argv))  
```

If I run this with two files (`1.input.file`, and `2.input.file`), the pipeline runs through fine and the required output files are created. 

```bash
$ ll
total 104
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:21 1.input.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 1.input.output1.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 1.input.output1.output2.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 1.input.output1.output2.output3.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:05 2.input.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 2.input.output1.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 2.input.output1.output2.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:23 2.input.output1.output2.output3.file
```

If I now touch the first input file, the pipeline detects that `1.input.output1.file` is out of date and updates. But if fails to update `1.input.output1.output2.file`, reporting that it is already up to date. Revlant part of the log:

```
2019-02-07 14:26:03,120 INFO Task enters queue = 'step1' 
# 2019-02-07 14:26:03,122 INFO {"task": "'step1'", "task_status": "update", "task_total": 2, "task_completed": 1, "task_completed_percent": 50.0}
# 2019-02-07 14:26:03,121 INFO     Job  = [1.input.file -> 1.input.output1.file]  \
#                                Input files: \
#                                 * 07 Feb 2019 14:25:58.00: 1.input.file \
#                                Output files: \
#                                 * 07 Feb 2019 14:22:20.00: 1.input.output1.file \
#                               
# 2019-02-07 14:26:03,223 INFO {"task": "'step1'", "task_status": "update", "task_total": 2, "task_completed": 1, "task_completed_percent": 50.0}
# 2019-02-07 14:26:03,223 INFO     Job  = [2.input.file -> 2.input.output1.file] unnecessary: already Up to date \
#                               \
#                                Input files: \
#                                   07 Feb 2019 14:05:20.00: 2.input.file \
#                                Output files: \
#                                   07 Feb 2019 14:22:20.00: 2.input.output1.file \
#                              
# 2019-02-07 14:26:03,327 INFO {"task": "'step1'", "task_status": "update", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,327 INFO     Job  = [1.input.file -> 1.input.output1.file] completed
# 2019-02-07 14:26:03,329 INFO {"task": "'step1'", "task_status": "completed", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,329 INFO Completed Task = 'step1' 
# 2019-02-07 14:26:03,329 INFO Task enters queue = 'step2' 
# 2019-02-07 14:26:03,330 INFO {"task": "'step2'", "task_status": "update", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,330 INFO     Job  = [1.input.output1.file -> 1.input.output1.output2.file] unnecessary: already Up to date \
#                               \
#                                Input files: \
#                                   07 Feb 2019 14:22:20.00: 1.input.output1.file \
#                                Output files: \
#                                   07 Feb 2019 14:22:31.00: 1.input.output1.output2.file \
#                              
# 2019-02-07 14:26:03,331 INFO {"task": "'step2'", "task_status": "update", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,331 INFO     Job  = [2.input.output1.file -> 2.input.output1.output2.file] unnecessary: already Up to date \
#                               \
#                                Input files: \
#                                   07 Feb 2019 14:22:20.00: 2.input.output1.file \
#                                Output files: \
#                                   07 Feb 2019 14:22:41.00: 2.input.output1.output2.file \
#                              
# 2019-02-07 14:26:03,332 INFO {"task": "'step2'", "task_status": "uptodate", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,332 INFO Uptodate Task = 'step2'
# 2019-02-07 14:26:03,333 INFO Task enters queue = 'step3' 
# 2019-02-07 14:26:03,334 INFO {"task": "'step3'", "task_status": "update", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,334 INFO     Job  = [1.input.output1.output2.file -> 1.input.output1.output2.output3.file] unnecessary: already Up to date \
#                               \
#                                Input files: \
#                                   07 Feb 2019 14:22:31.00: 1.input.output1.output2.file \
#                                Output files: \
#                                   07 Feb 2019 14:22:51.00: 1.input.output1.output2.output3.file \
#                              
# 2019-02-07 14:26:03,335 INFO {"task": "'step3'", "task_status": "update", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,335 INFO     Job  = [2.input.output1.output2.file -> 2.input.output1.output2.output3.file] unnecessary: already Up to date \
#                               \
#                                Input files: \
#                                   07 Feb 2019 14:22:41.00: 2.input.output1.output2.file \
#                                Output files: \
#                                   07 Feb 2019 14:23:01.00: 2.input.output1.output2.output3.file \
#                              
# 2019-02-07 14:26:03,336 INFO {"task": "'step3'", "task_status": "uptodate", "task_total": 2, "task_completed": 2, "task_completed_percent": 100.0}
# 2019-02-07 14:26:03,335 INFO Uptodate Task = 'step3'
# 2019-02-07 14:26:03,349 INFO job finished in 0 seconds at Thu Feb  7 14:26:03 2019 --  1.12  0.63  0.01  0.02 -- a92c63a6-a4df-4ece-944b-162bdfab2af4
```

If we look now, `1.input.output1.file` has the correct timestamp, but `ruffus` appears to have failed to pick it up. 

```bash
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:25 1.input.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:26 1.input.output1.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 1.input.output1.output2.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 1.input.output1.output2.output3.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:05 2.input.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 2.input.output1.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:22 2.input.output1.output2.file
-rw-rw-r-- 1 UUUU GG      0 Feb  7 14:23 2.input.output1.output2.output3.file
```

**Comment by AndreasHeger** [comment]
This is a cgatcore issue/feature.
In cgatcore, all the os.stat calls are cached to prevent stat'ing the same file multiple times. The reason for this is that this is a slow operation on networked file systems.
